File: Algorithms/CP/Ruby/cp.rb
#!/usr/bin/env ruby
# CP (Constraint Propagation) Sudoku Solver - Ruby Implementation
# Uses constraint propagation with MRV (Minimum Remaining Values) heuristic
#
# Algorithm:
# - Track candidates for each cell using bitsets
# - Propagate constraints (singleton elimination + hidden singles)
# - Use MRV heuristic for cell selection during search
# - Count iterations via assign() calls

# Global iteration counter
$cp_iterations = 0

# CPGrid class - holds values and candidates
class CPGrid
  attr_accessor :values, :candidates

  def initialize
    @values = Array.new(9) { Array.new(9, 0) }
    @candidates = Array.new(9) { Array.new(9, 0) }
  end

  def deep_copy
    copy = CPGrid.new
    9.times do |r|
      9.times do |c|
        copy.values[r][c] = @values[r][c]
        copy.candidates[r][c] = @candidates[r][c]
      end
    end
    copy
  end
end

# Bitset helper methods
def has_candidate?(set, digit)
  (set & (1 << digit)) != 0
end

def add_candidate(set, digit)
  set | (1 << digit)
end

def remove_candidate(set, digit)
  set & ~(1 << digit)
end

def count_candidates(set)
  # Count set bits using Ruby's built-in method
  set.to_s(2).count('1')
end

def get_first_candidate(set)
  (1..9).each do |digit|
    return digit if has_candidate?(set, digit)
  end
  0
end

# Get all 20 peers for a cell (row, col, box)
def get_peers(row, col)
  peers = []

  # Same row (8 cells)
  9.times do |c|
    peers << [row, c] if c != col
  end

  # Same column (8 cells)
  9.times do |r|
    peers << [r, col] if r != row
  end

  # Same 3x3 box (4 remaining cells)
  box_row = (row / 3) * 3
  box_col = (col / 3) * 3
  (box_row...box_row + 3).each do |r|
    (box_col...box_col + 3).each do |c|
      peers << [r, c] if r != row && c != col
    end
  end

  peers
end

# Initialize grid with puzzle
def init_grid(grid, puzzle)
  9.times do |row|
    9.times do |col|
      if puzzle[row][col] == 0
        # Empty cell: set all candidates 1-9
        grid.values[row][col] = 0
        grid.candidates[row][col] = 0x3FE  # Binary: 0011 1111 1110 (bits 1-9)
      else
        # Given clue: set single value
        digit = puzzle[row][col]
        grid.values[row][col] = digit
        grid.candidates[row][col] = (1 << digit)
      end
    end
  end
end

# Eliminate a digit from a cell's candidates
def eliminate(grid, row, col, digit)
  # Check if digit is already eliminated
  return true unless has_candidate?(grid.candidates[row][col], digit)

  # Remove digit from candidates
  grid.candidates[row][col] = remove_candidate(grid.candidates[row][col], digit)

  # Check for contradiction (no candidates left)
  remaining = count_candidates(grid.candidates[row][col])
  return false if remaining == 0

  # If only one candidate left, assign it (singleton elimination)
  if remaining == 1 && grid.values[row][col] == 0
    last_digit = get_first_candidate(grid.candidates[row][col])
    return false unless assign(grid, row, col, last_digit)
  end

  true
end

# Assign a value to a cell
def assign(grid, row, col, digit)
  # Increment iteration counter
  $cp_iterations += 1

  # Set value
  grid.values[row][col] = digit
  grid.candidates[row][col] = (1 << digit)

  # Eliminate digit from all peers
  peers = get_peers(row, col)
  peers.each do |peer_row, peer_col|
    return false unless eliminate(grid, peer_row, peer_col, digit)
  end

  true
end

# Propagate constraints until quiescence
def propagate(grid)
  changed = true

  while changed
    changed = false

    # Strategy 1: Singleton elimination
    # If a cell has only one candidate, assign it
    9.times do |row|
      9.times do |col|
        next if grid.values[row][col] != 0

        num_candidates = count_candidates(grid.candidates[row][col])
        return false if num_candidates == 0

        if num_candidates == 1
          digit = get_first_candidate(grid.candidates[row][col])
          return false unless assign(grid, row, col, digit)
          changed = true
        end
      end
    end

    # Strategy 2: Hidden singles - rows
    9.times do |row|
      (1..9).each do |digit|
        count = 0
        last_col = -1

        9.times do |col|
          if grid.values[row][col] == digit
            count = 0  # Already assigned
            break
          end
          if has_candidate?(grid.candidates[row][col], digit)
            count += 1
            last_col = col
          end
        end

        if count == 1
          return false unless assign(grid, row, last_col, digit)
          changed = true
        elsif count == 0
          # Check if digit is already assigned in this row
          found = (0...9).any? { |col| grid.values[row][col] == digit }
          return false unless found
        end
      end
    end

    # Strategy 2: Hidden singles - columns
    9.times do |col|
      (1..9).each do |digit|
        count = 0
        last_row = -1

        9.times do |row|
          if grid.values[row][col] == digit
            count = 0  # Already assigned
            break
          end
          if has_candidate?(grid.candidates[row][col], digit)
            count += 1
            last_row = row
          end
        end

        if count == 1
          return false unless assign(grid, last_row, col, digit)
          changed = true
        elsif count == 0
          # Check if digit is already assigned in this column
          found = (0...9).any? { |row| grid.values[row][col] == digit }
          return false unless found
        end
      end
    end

    # Strategy 2: Hidden singles - boxes
    9.times do |box|
      box_row = (box / 3) * 3
      box_col = (box % 3) * 3

      (1..9).each do |digit|
        count = 0
        last_r = -1
        last_c = -1
        already_assigned = false

        (box_row...box_row + 3).each do |r|
          (box_col...box_col + 3).each do |c|
            if grid.values[r][c] == digit
              already_assigned = true
              break
            end
            if has_candidate?(grid.candidates[r][c], digit)
              count += 1
              last_r = r
              last_c = c
            end
          end
          break if already_assigned
        end

        next if already_assigned

        if count == 1
          return false unless assign(grid, last_r, last_c, digit)
          changed = true
        elsif count == 0
          # Check if digit is already assigned in this box
          found = false
          (box_row...box_row + 3).each do |r|
            (box_col...box_col + 3).each do |c|
              if grid.values[r][c] == digit
                found = true
                break
              end
            end
            break if found
          end
          return false unless found
        end
      end
    end
  end

  true  # Success - reached fixpoint
end

# Find cell with Minimum Remaining Values
def find_mrv_cell(grid)
  min_candidates = 10
  mrv_row = -1
  mrv_col = -1

  9.times do |r|
    9.times do |c|
      next if grid.values[r][c] != 0

      num_candidates = count_candidates(grid.candidates[r][c])
      if num_candidates < min_candidates
        min_candidates = num_candidates
        mrv_row = r
        mrv_col = c
      end
    end
  end

  return nil if mrv_row == -1  # No empty cells

  [mrv_row, mrv_col]
end

# CP Search with backtracking
def cp_search(grid, solution)
  # Base case: check if grid is complete
  mrv_cell = find_mrv_cell(grid)

  unless mrv_cell
    # No empty cells - grid is complete
    9.times do |r|
      9.times do |c|
        solution[r * 9 + c] = grid.values[r][c]
      end
    end
    return true
  end

  mrv_row, mrv_col = mrv_cell
  candidates = grid.candidates[mrv_row][mrv_col]

  # Try each candidate for the MRV cell
  (1..9).each do |digit|
    next unless has_candidate?(candidates, digit)

    # Save grid state for backtracking
    grid_copy = grid.deep_copy

    # Try assigning this digit
    if assign(grid, mrv_row, mrv_col, digit)
      # Assignment succeeded, propagate constraints
      if propagate(grid)
        # Propagation succeeded, recurse
        return true if cp_search(grid, solution)
      end
    end

    # Failed - restore grid state and try next candidate
    grid.values = grid_copy.values
    grid.candidates = grid_copy.candidates
  end

  # All candidates exhausted - dead end
  false
end

# Print puzzle
def print_puzzle(grid)
  puts "\nPuzzle:"
  grid.each do |row|
    puts row.map(&:to_s).join(" ") + " "
  end
end

# Read matrix file
def read_matrix_file(filename)
  # Normalize path for output
  display_path = filename
  if filename.start_with?("/app/Matrices/")
    display_path = "../" + filename[5..-1]
  end
  puts display_path

  puzzle = Array.new(9) { Array.new(9, 0) }
  line_count = 0

  File.foreach(filename) do |line|
    line = line.strip
    next if line.empty? || line.start_with?('#')

    values = line.split.map(&:to_i)
    if values.length == 9 && line_count < 9
      puzzle[line_count] = values
      puts values.map(&:to_s).join(" ") + " "
      line_count += 1
    end
  end

  puzzle
end

# Main program
if ARGV.empty?
  puts "Usage: ruby cp.rb <matrix_file>"
  exit 1
end

start_time = Time.now

ARGV.each do |filename|
  next unless filename.end_with?(".matrix")

  # Read puzzle from file
  puzzle = read_matrix_file(filename)
  print_puzzle(puzzle)

  # Initialize CP grid
  grid = CPGrid.new
  init_grid(grid, puzzle)

  # Apply initial propagation
  unless propagate(grid)
    puts "\nNo solution found (contradiction during initial propagation)"
    elapsed = Time.now - start_time
    puts "Seconds to process %.3f" % elapsed
    next
  end

  # Run search
  $cp_iterations = 0
  solution = Array.new(81)
  solved = cp_search(grid, solution)

  if solved
    # Convert solution array back to 2D for printing
    solution_grid = Array.new(9) { Array.new(9, 0) }
    9.times do |r|
      9.times do |c|
        solution_grid[r][c] = solution[r * 9 + c]
      end
    end

    print_puzzle(solution_grid)
    puts "\nSolved in Iterations=#{$cp_iterations}\n"
  else
    puts "\nNo solution found"
  end
end

elapsed = Time.now - start_time
puts "Seconds to process %.3f" % elapsed

File: Algorithms/CP/Perl/cp.pl
#!/usr/bin/env perl
# Constraint Propagation (CP) Sudoku Solver - Perl Implementation
# Mechanical translation from C reference to preserve algorithm correctness.
#
# Algorithm: Constraint propagation with MRV (Minimum Remaining Values) heuristic
# - Uses bitsets to track candidate values (bits 1-9)
# - Propagates constraints: singleton elimination, hidden singles
# - Search with MRV cell selection for efficiency

use strict;
use warnings;
use Time::HiRes qw(time);

# Global iteration counter
my $cp_iterations = 0;

# CPGrid structure (hash reference)
sub new_cp_grid {
    return {
        values => [map { [(0) x 9] } (0..8)],         # Assigned values (0 = empty)
        candidates => [map { [(0) x 9] } (0..8)]      # Possible values per cell (bitset)
    };
}

# Bitset helper functions
sub has_candidate {
    my ($candidate_set, $digit) = @_;
    return ($candidate_set & (1 << $digit)) != 0;
}

sub add_candidate {
    my ($candidate_set, $digit) = @_;
    return $candidate_set | (1 << $digit);
}

sub remove_candidate {
    my ($candidate_set, $digit) = @_;
    return $candidate_set & ~(1 << $digit);
}

sub count_candidates {
    my ($candidate_set) = @_;
    my $count = 0;
    while ($candidate_set) {
        $count += $candidate_set & 1;
        $candidate_set >>= 1;
    }
    return $count;
}

sub get_first_candidate {
    my ($candidate_set) = @_;
    for my $digit (1..9) {
        return $digit if has_candidate($candidate_set, $digit);
    }
    return 0;
}

# Get all 20 peers for a cell (row, col, box)
sub get_peers {
    my ($row, $col) = @_;
    my @peers;

    # Same row (9 cells minus self = 8)
    for my $c (0..8) {
        push @peers, [$row, $c] if $c != $col;
    }

    # Same column (9 cells minus self = 8)
    for my $r (0..8) {
        push @peers, [$r, $col] if $r != $row;
    }

    # Same 3x3 box (9 cells minus self minus already counted = 4)
    my $box_row = int($row / 3) * 3;
    my $box_col = int($col / 3) * 3;
    for my $r ($box_row .. $box_row + 2) {
        for my $c ($box_col .. $box_col + 2) {
            push @peers, [$r, $c] if $r != $row && $c != $col;
        }
    }

    return @peers;
}

# Initialize grid from puzzle
sub init_grid {
    my ($grid, $puzzle) = @_;
    for my $row (0..8) {
        for my $col (0..8) {
            if ($puzzle->[$row][$col] == 0) {
                # Empty cell: set all candidates 1-9 (bits 1-9 set)
                $grid->{values}[$row][$col] = 0;
                $grid->{candidates}[$row][$col] = 0x3FE;  # Binary: 0011 1111 1110
            } else {
                # Given clue: set single value
                my $digit = $puzzle->[$row][$col];
                $grid->{values}[$row][$col] = $digit;
                $grid->{candidates}[$row][$col] = (1 << $digit);
            }
        }
    }
}

# Forward declaration for mutual recursion
sub assign;

# Eliminate digit from candidates at (row, col)
sub eliminate {
    my ($grid, $row, $col, $digit) = @_;

    # Check if digit is already eliminated
    return 1 unless has_candidate($grid->{candidates}[$row][$col], $digit);

    # Remove digit from candidates
    $grid->{candidates}[$row][$col] = remove_candidate($grid->{candidates}[$row][$col], $digit);

    # Check for contradiction (no candidates left)
    my $remaining = count_candidates($grid->{candidates}[$row][$col]);
    return 0 if $remaining == 0;

    # If only one candidate left, assign it (singleton elimination)
    if ($remaining == 1 && $grid->{values}[$row][$col] == 0) {
        my $last_digit = get_first_candidate($grid->{candidates}[$row][$col]);
        return 0 unless assign($grid, $row, $col, $last_digit);
    }

    return 1;
}

# Assign digit to cell at (row, col)
sub assign {
    my ($grid, $row, $col, $digit) = @_;

    # Increment iteration counter (this is our benchmark metric)
    $cp_iterations++;

    # Set value
    $grid->{values}[$row][$col] = $digit;
    $grid->{candidates}[$row][$col] = (1 << $digit);

    # Eliminate digit from all peers
    my @peers = get_peers($row, $col);
    for my $peer (@peers) {
        my ($peer_row, $peer_col) = @$peer;
        return 0 unless eliminate($grid, $peer_row, $peer_col, $digit);
    }

    return 1;
}

# Apply constraint propagation until quiescence
sub propagate {
    my ($grid) = @_;
    my $changed = 1;

    while ($changed) {
        $changed = 0;

        # Strategy 1: Singleton elimination
        # If a cell has only one candidate, assign it
        for my $row (0..8) {
            for my $col (0..8) {
                if ($grid->{values}[$row][$col] == 0) {
                    my $num_candidates = count_candidates($grid->{candidates}[$row][$col]);
                    return 0 if $num_candidates == 0;  # Contradiction
                    if ($num_candidates == 1) {
                        my $digit = get_first_candidate($grid->{candidates}[$row][$col]);
                        return 0 unless assign($grid, $row, $col, $digit);
                        $changed = 1;
                    }
                }
            }
        }

        # Strategy 2: Hidden singles
        # For each unit (row, col, box), if a digit appears in only one cell, assign it

        # Check rows
        for my $row (0..8) {
            for my $digit (1..9) {
                my $count = 0;
                my $last_col = -1;
                my $already_assigned = 0;

                for my $col (0..8) {
                    if ($grid->{values}[$row][$col] == $digit) {
                        $already_assigned = 1;
                        last;
                    }
                    if (has_candidate($grid->{candidates}[$row][$col], $digit)) {
                        $count++;
                        $last_col = $col;
                    }
                }

                next if $already_assigned;

                if ($count == 1) {
                    return 0 unless assign($grid, $row, $last_col, $digit);
                    $changed = 1;
                } elsif ($count == 0) {
                    return 0;  # Digit cannot be placed anywhere in row
                }
            }
        }

        # Check columns
        for my $col (0..8) {
            for my $digit (1..9) {
                my $count = 0;
                my $last_row = -1;
                my $already_assigned = 0;

                for my $row (0..8) {
                    if ($grid->{values}[$row][$col] == $digit) {
                        $already_assigned = 1;
                        last;
                    }
                    if (has_candidate($grid->{candidates}[$row][$col], $digit)) {
                        $count++;
                        $last_row = $row;
                    }
                }

                next if $already_assigned;

                if ($count == 1) {
                    return 0 unless assign($grid, $last_row, $col, $digit);
                    $changed = 1;
                } elsif ($count == 0) {
                    return 0;  # Digit cannot be placed anywhere in column
                }
            }
        }

        # Check boxes
        for my $box (0..8) {
            my $box_row = int($box / 3) * 3;
            my $box_col = ($box % 3) * 3;

            for my $digit (1..9) {
                my $count = 0;
                my ($last_r, $last_c) = (-1, -1);
                my $already_assigned = 0;

                FIND_IN_BOX: for my $r ($box_row .. $box_row + 2) {
                    for my $c ($box_col .. $box_col + 2) {
                        if ($grid->{values}[$r][$c] == $digit) {
                            $already_assigned = 1;
                            last FIND_IN_BOX;
                        }
                        if (has_candidate($grid->{candidates}[$r][$c], $digit)) {
                            $count++;
                            ($last_r, $last_c) = ($r, $c);
                        }
                    }
                }

                next if $already_assigned;

                if ($count == 1) {
                    return 0 unless assign($grid, $last_r, $last_c, $digit);
                    $changed = 1;
                } elsif ($count == 0) {
                    return 0;  # Digit cannot be placed anywhere in box
                }
            }
        }
    }

    return 1;  # Success - reached fixpoint
}

# Find empty cell with Minimum Remaining Values (fewest candidates)
sub find_mrv_cell {
    my ($grid) = @_;
    my $min_candidates = 10;  # More than 9
    my ($mrv_row, $mrv_col) = (-1, -1);
    my $found = 0;

    for my $r (0..8) {
        for my $c (0..8) {
            if ($grid->{values}[$r][$c] == 0) {
                my $num_candidates = count_candidates($grid->{candidates}[$r][$c]);
                if ($num_candidates < $min_candidates) {
                    $min_candidates = $num_candidates;
                    ($mrv_row, $mrv_col) = ($r, $c);
                    $found = 1;
                }
            }
        }
    }

    return $found ? ($mrv_row, $mrv_col) : ();
}

# Deep copy grid for backtracking
sub copy_grid {
    my ($grid) = @_;
    my $copy = new_cp_grid();
    for my $r (0..8) {
        for my $c (0..8) {
            $copy->{values}[$r][$c] = $grid->{values}[$r][$c];
            $copy->{candidates}[$r][$c] = $grid->{candidates}[$r][$c];
        }
    }
    return $copy;
}

# Search with constraint propagation
sub cp_search {
    my ($grid, $solution) = @_;

    # Base case: check if grid is complete
    my @mrv_cell = find_mrv_cell($grid);
    unless (@mrv_cell) {
        # No empty cells - grid is complete, extract solution
        for my $r (0..8) {
            for my $c (0..8) {
                $solution->[$r * 9 + $c] = $grid->{values}[$r][$c];
            }
        }
        return 1;
    }

    my ($mrv_row, $mrv_col) = @mrv_cell;

    # Recursive case: try each candidate for the MRV cell
    my $candidates = $grid->{candidates}[$mrv_row][$mrv_col];

    for my $digit (1..9) {
        if (has_candidate($candidates, $digit)) {
            # Save grid state for backtracking
            my $grid_copy = copy_grid($grid);

            # Try assigning this digit
            if (assign($grid, $mrv_row, $mrv_col, $digit)) {
                # Assignment succeeded, propagate constraints
                if (propagate($grid)) {
                    # Propagation succeeded, recurse
                    return 1 if cp_search($grid, $solution);
                }
            }

            # Failed - restore grid state and try next candidate
            for my $r (0..8) {
                for my $c (0..8) {
                    $grid->{values}[$r][$c] = $grid_copy->{values}[$r][$c];
                    $grid->{candidates}[$r][$c] = $grid_copy->{candidates}[$r][$c];
                }
            }
        }
    }

    # All candidates exhausted - dead end
    return 0;
}

# Print puzzle
sub print_puzzle {
    my ($puzzle) = @_;
    print "\nPuzzle:\n";
    for my $r (0..8) {
        print join(" ", @{$puzzle->[$r]}) . " \n";
    }
}

# Read matrix file
sub read_matrix_file {
    my ($filename) = @_;
    my @puzzle = map { [(0) x 9] } (0..8);

    # Normalize path for output
    my $display_path = $filename;
    if ($filename =~ /^\/app\/Matrices\//) {
        $display_path = "../" . substr($filename, 5);
    }
    print "$display_path\n";

    open(my $fh, '<', $filename) or die "Cannot open $filename: $!";
    my $line_count = 0;

    while (my $line = <$fh>) {
        chomp $line;
        $line =~ s/^\s+|\s+$//g;

        # Skip comments and empty lines
        next if $line eq '' || $line =~ /^#/;

        # Parse 9 integers from line
        my @values = split(/\s+/, $line);
        if (scalar(@values) == 9 && $line_count < 9) {
            $puzzle[$line_count] = [@values];
            print join(" ", @values) . " \n";
            $line_count++;
        }
    }
    close($fh);

    die "Expected 9 lines, got $line_count" if $line_count != 9;
    return \@puzzle;
}

# Main program
unless (@ARGV == 1) {
    print STDERR "Usage: $0 <matrix_file>\n";
    exit 1;
}

my $start_time = time();

eval {
    # Read puzzle from file
    my $puzzle = read_matrix_file($ARGV[0]);
    print_puzzle($puzzle);

    # Initialize CP grid
    my $grid = new_cp_grid();
    init_grid($grid, $puzzle);

    # Apply initial propagation
    $cp_iterations = 0;
    unless (propagate($grid)) {
        print "\nNo solution found (contradiction during initial propagation)\n\n";
        my $elapsed = time() - $start_time;
        printf "Seconds to process %.3f\n", $elapsed;
        exit 0;
    }

    # Run search
    my @solution = (0) x 81;
    my $solved = cp_search($grid, \@solution);

    if ($solved) {
        # Convert solution array back to 2D for printing
        my @solution_grid = map { [(0) x 9] } (0..8);
        for my $r (0..8) {
            for my $c (0..8) {
                $solution_grid[$r][$c] = $solution[$r * 9 + $c];
            }
        }

        print_puzzle(\@solution_grid);
        print "\nSolved in Iterations=$cp_iterations\n\n";
    } else {
        print "\nNo solution found\n\n";
    }
};
if ($@) {
    print STDERR "Error: $@\n";
    exit 1;
}

my $elapsed = time() - $start_time;
printf "Seconds to process %.3f\n", $elapsed;
